File: bitcoin_stockmem/evaluation/backtest.py
"""Rolling-window backtesting with online learning.

Follows the paper's evaluation protocol:
  - Train on period A (build event + reflection memory).
  - Test on period B, day-by-day in chronological order.
  - After each test prediction, add ground truth + reflection (online learning).
"""

from __future__ import annotations

import logging
from collections import defaultdict

import pandas as pd

from config import ASSETS, WINDOW_SIZE
from evaluation.metrics import evaluate
from llm.gemini_client import GeminiClient
from memory.event_memory import compute_and_store_daily_vectors
from pipeline.step1_extract import extract_events_for_day
from pipeline.step2_merge import merge_events_for_day
from pipeline.step3_track import track_events_for_day
from pipeline.step4_reason import generate_reflection
from pipeline.step6_predict import predict
from storage.database import get_cursor, insert_raw_news

logger = logging.getLogger(__name__)


def _group_news_by_date(news: list[dict]) -> dict[str, list[dict]]:
    groups = defaultdict(list)
    for article in news:
        groups[article["date"]].append(article)
    return dict(groups)


def build_event_memory(
    client: GeminiClient,
    news_by_date: dict[str, list[dict]],
    dates: list[str],
) -> None:
    """Run steps 1-3 (extract, merge, track) for a list of dates."""
    for date in sorted(dates):
        articles = news_by_date.get(date, [])

        # Store news
        if articles:
            insert_raw_news(articles)

        # Step 1: Extract
        events = extract_events_for_day(client, articles, date)
        if events:
            from storage.database import insert_raw_events
            insert_raw_events(events)

        # Step 2: Merge
        merge_events_for_day(client, date)

        # Step 3: Track
        track_events_for_day(client, date)

        # Compute daily vectors for all assets
        for asset in ASSETS:
            compute_and_store_daily_vectors(date, asset)

        logger.info(f"Event memory built for {date}")


def build_reflection_memory(
    client: GeminiClient,
    labels: dict[str, pd.DataFrame],
) -> None:
    """Run step 4 (reflection) for all training labels.

    Args:
        labels: {asset: DataFrame with 'date', 'label', 'next_return'}.
    """
    for asset, df in labels.items():
        for _, row in df.iterrows():
            if row["label"] == "flat":
                continue
            generate_reflection(
                client,
                date=row["date"],
                asset=asset,
                price_direction=row["label"],
                price_change_pct=row.get("next_return"),
                source="train",
            )
        logger.info(f"Reflections built for {asset} training data")


def run_backtest(
    client: GeminiClient,
    test_news_by_date: dict[str, list[dict]],
    test_labels: dict[str, pd.DataFrame],
    test_dates: list[str],
) -> dict[str, dict]:
    """Run the full backtest loop on test data.

    Args:
        client: Gemini client.
        test_news_by_date: {date: [articles]} for test period.
        test_labels: {asset: DataFrame} with test labels.
        test_dates: Sorted list of test dates.

    Returns:
        {asset: {accuracy, mcc, total, correct, predictions}}
    """
    results: dict[str, dict] = {}
    predictions_log: dict[str, list[tuple[str, str]]] = defaultdict(list)

    for date in test_dates:
        # Build event memory for this day
        articles = test_news_by_date.get(date, [])
        if articles:
            insert_raw_news(articles)
            events = extract_events_for_day(client, articles, date)
            if events:
                from storage.database import insert_raw_events
                insert_raw_events(events)

        merge_events_for_day(client, date)
        track_events_for_day(client, date)

        for asset in ASSETS:
            compute_and_store_daily_vectors(date, asset)

        # Predict for each asset
        for asset in ASSETS:
            df = test_labels.get(asset)
            if df is None:
                continue

            day_row = df[df["date"] == date]
            if day_row.empty:
                continue

            actual = day_row.iloc[0]["label"]
            if actual == "flat":
                continue

            # Step 5-6: Retrieve + Predict
            pred = predict(client, date, asset)
            if not pred:
                continue

            predicted = pred["predicted_direction"]
            predictions_log[asset].append((predicted, actual))

            # Update prediction with actual
            with get_cursor() as cur:
                cur.execute(
                    "UPDATE predictions SET actual_direction = ? WHERE id = ?",
                    (actual, pred["id"]),
                )

            # Online learning: generate reflection with true label
            generate_reflection(
                client,
                date=date,
                asset=asset,
                price_direction=actual,
                price_change_pct=day_row.iloc[0].get("next_return"),
                source="online",
            )

            logger.info(
                f"Test {date}/{asset}: predicted={predicted}, actual={actual}"
            )

    # Compute metrics
    for asset in ASSETS:
        preds_actuals = predictions_log.get(asset, [])
        if preds_actuals:
            preds = [p for p, _ in preds_actuals]
            actuals = [a for _, a in preds_actuals]
            metrics = evaluate(preds, actuals)
            metrics["predictions"] = preds_actuals
            results[asset] = metrics
            logger.info(f"{asset} results: ACC={metrics['accuracy']}, MCC={metrics['mcc']}")

    return results

File: bitcoin-stockmem-ts/src/evaluation/backtest.ts
/**
 * Rolling-window backtesting with online learning.
 */

import { ASSETS, type Asset } from "../config";
import { GeminiClient } from "../llm/gemini-client";
import { evaluate } from "./metrics";
import { insertRawNews, insertRawEvents, getDb } from "../storage/database";
import { computeAndStoreDailyVectors } from "../memory/event-memory";
import { extractEventsForDay } from "../pipeline/step1-extract";
import { mergeEventsForDay } from "../pipeline/step2-merge";
import { trackEventsForDay } from "../pipeline/step3-track";
import { generateReflection } from "../pipeline/step4-reason";
import { predict } from "../pipeline/step6-predict";
import type { NewsArticle } from "../data/news-fetcher";
import type { LabelledRow } from "../data/label-generator";

export async function buildEventMemory(
  client: GeminiClient,
  newsByDate: Record<string, NewsArticle[]>,
  dates: string[]
): Promise<void> {
  for (const date of dates.sort()) {
    const articles = newsByDate[date] ?? [];

    if (articles.length) {
      insertRawNews(articles.map((a) => ({
        date: a.date, source: a.source, title: a.title,
        body: a.body, url: a.url, asset: a.asset,
      })));
    }

    const events = await extractEventsForDay(client, articles, date);
    if (events.length) insertRawEvents(events);

    await mergeEventsForDay(client, date);
    await trackEventsForDay(client, date);

    for (const asset of ASSETS) {
      computeAndStoreDailyVectors(date, asset);
    }
    console.log(`Event memory built for ${date}`);
  }
}

export async function buildReflectionMemory(
  client: GeminiClient,
  labels: Record<string, LabelledRow[]>
): Promise<void> {
  for (const [asset, rows] of Object.entries(labels)) {
    for (const row of rows) {
      if (row.label === "flat") continue;
      await generateReflection(
        client, row.date, asset, row.label, row.next_return ?? undefined, "train"
      );
    }
    console.log(`Reflections built for ${asset} training data`);
  }
}

export async function runBacktest(
  client: GeminiClient,
  testNewsByDate: Record<string, NewsArticle[]>,
  testLabels: Record<string, LabelledRow[]>,
  testDates: string[]
): Promise<Record<string, ReturnType<typeof evaluate> & { predictions: [string, string][] }>> {
  const results: Record<string, any> = {};
  const predLog: Record<string, [string, string][]> = {};
  for (const a of ASSETS) predLog[a] = [];

  for (const date of testDates) {
    const articles = testNewsByDate[date] ?? [];
    if (articles.length) {
      insertRawNews(articles.map((a) => ({
        date: a.date, source: a.source, title: a.title,
        body: a.body, url: a.url, asset: a.asset,
      })));
    }

    const events = await extractEventsForDay(client, articles, date);
    if (events.length) insertRawEvents(events);
    await mergeEventsForDay(client, date);
    await trackEventsForDay(client, date);
    for (const asset of ASSETS) computeAndStoreDailyVectors(date, asset);

    for (const asset of ASSETS) {
      const rows = testLabels[asset] ?? [];
      const dayRow = rows.find((r) => r.date === date);
      if (!dayRow || dayRow.label === "flat") continue;

      const pred = await predict(client, date, asset);
      if (!pred) continue;

      predLog[asset].push([pred.predictedDirection, dayRow.label]);

      // Update actual
      const d = getDb();
      d.prepare("UPDATE predictions SET actual_direction = ? WHERE id = ?")
        .run(dayRow.label, pred.id);

      // Online learning
      await generateReflection(
        client, date, asset, dayRow.label, dayRow.next_return ?? undefined, "online"
      );

      console.log(`Test ${date}/${asset}: predicted=${pred.predictedDirection}, actual=${dayRow.label}`);
    }
  }

  for (const asset of ASSETS) {
    const pairs = predLog[asset];
    if (pairs.length) {
      const preds = pairs.map(([p]) => p);
      const actuals = pairs.map(([, a]) => a);
      results[asset] = { ...evaluate(preds, actuals), predictions: pairs };
      console.log(`${asset}: ACC=${results[asset].accuracy}, MCC=${results[asset].mcc}`);
    }
  }
  return results;
}
